Products/Zuul/catalog/reindex_solr.py

```diff
@@ -0,0 +1,308 @@
+            return WorkerReport(self.idx, count)
+        except Queue.Empty:
+            # The parent_queue is empty, which means we're done
+            self.index(batch, True)
+            return WorkerReport(self.idx, count)
+        except Exception as e:
+            # grab the traceback for the report
+            return WorkerReport(self.idx, count - len(batch), traceback.format_exc())
+
+    def run(self):
+        self.queue.put(self.do_work())
+
+
+class HardReindex(ReindexProcess):
+    """
+    A ReindexProcess which uses a tree splitting strategy to share work.
+
+    General flow for a single HardReindex process is:
+    If deque is empty, get node from parent queue.
+    Else get node from right side of deque.
+    Push node's children to right side of deque (if applicable).
+    Add node to list of items to be indexed.
+    Index all items in list if fit.
+    If parent queue is empty, pop a node from left side of deque and push to parent queue.
+    Repeat.
+    """
+    def __init__(self, queue, idx, parent_queue):
+        super(HardReindex, self).__init__(queue, idx, parent_queue)
+        self.deque = deque()
+
+    def get_work(self):
+        yield self.parent_queue.get(timeout=30)
+        while True:
+            try:
+                yield self.deque.pop()
+            except IndexError:
+                yield self.parent_queue.get(timeout=2)
+
+    def push_children(self, node):
+        if self._include_children(node):
+            prefix = "/".join(node.getPhysicalPath()) + "/"
+            [self.deque.append(prefix + child) for child in node.objectIds()]
+
+    def _include_children(self, node):
+        include = False
+        if not isinstance(node, GlobalCatalog):
+            if isinstance(node, ObjectManager):
+                include = True
+            elif isinstance(node, ToManyContRelationship):
+                include = True
+        return include
+
+    def include_node(self, node):
+        include = False
+        if not isinstance(node, GlobalCatalog):
+            if isinstance(node,ZenModelRM):
+                include = True
+        return include
+
+    def index(self, batch, commit=False):
+        self.batch_update(batch, commit)
+
+    def reconcile(self):
+        try:
+            if self.parent_queue.empty():
+                self.parent_queue.put(self.deque.popleft())
+        except IndexError:
+            pass
+
+
+class SoftReindex(ReindexProcess):
+    """
+    A ReindexProcess which uses a workerpool style strategy to work.
+
+    General flow for a single SoftReindex process is:
+    Get node from parent queue.
+    Add node to list of items to be indexed.
+    Index all items in list if fit.
+    Repeat.
+    """
+    def __init__(self, queue, idx, parent_queue):
+        super(SoftReindex, self).__init__(queue, idx, parent_queue)
+
+    def get_work(self):
+        yield self.parent_queue.get(timeout=30)
+        while True:
+            yield self.parent_queue.get(timeout=2)
```

Comment:
add a comment to clarify the method raises QueueEmpty excpetion or add a try/except and return empy list when there are no more elements in the queue. Same for the hard version